[Docs] Add declarative peering section (#2813)
liqotech/liqo: docs/advanced/peering/peering-via-cr.md

# Declarative peering

Declarative peering is supported starting from Liqo 1.0, which means that you can create a set of CRs describing the peering with a remote cluster, which is automatically set up once the CRs are applied on both sides.
This simplifies automation, GitOps and continuous delivery: for example, you might have your Git repository with the manifests describing the peerings, and an instance of [ArgoCD](https://argo-cd.readthedocs.io) which synchronizes the changes on the clusters, creating and destroying the peerings.

> The following lines analizes how to describe the configuration of each of the Liqo modules:

Reviewer: I did not understand what "the following lines" means.

# Declarative peering

Declarative peerings are supported starting from Liqo 1.0, which means that you can create a set of CRs describing the peering with a remote cluster, which is automatically set up once the CRs are applied on both sides.
This simplifies automation, GitOps and continuous delivery: for example, you might have your Git repository with the manifests describing the peerings, and an instance of [ArgoCD](https://argo-cd.readthedocs.io) which synchronizes the changes on the clusters, creating and destroying the peerings.

The following lines analizes how to describe the configuration of each of the Liqo modules:

- Networking
- Authentication
- Offloading

## Tenant namespace

Before starting to configure the peerings, you need to create a tenant namespace in both clusters that you need to peer.
This tenant namespace must **refer to the peering with a specific cluster**, hence a distinct tenant namespace must be created per each peering.
**All the resources needed to configure the peering must be created in those namespaces**.
A tenant namespace **can have an arbitrary name**, but **it must have the following labels**:

```text

> liqo.io/remote-cluster-id: <HERE_THE_CLUSTER_ID_OF_PEER_CLUSTER>

Reviewer: I would add something like
where the remote-cluster-id can be obtained by.... (and write the command) (or)
where the remote-cluster-id can be something like... (and write a possible clusterid)

liqo.io/tenant-namespace: "true"
```

> ### Configuring the tenant namespace on cluster consumer

Reviewer: I would say "consumer cluster", "provider cluster", instead of "cluster consumer" and "cluster provider".

> The following is an example of tenant namespace referring to the peering with a cluster with id `cl-provider`:

Reviewer: ```suggestion
The following is an example of tenant namespace named `liqo-tenant-cl-provider`, which refers to the peering with a cluster with id `cl-provider`:
```

```yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    liqo.io/remote-cluster-id: cl-provider
    liqo.io/tenant-namespace: "true"
  name: liqo-tenant-cl-provider
spec: {}
```

### Configuring the tenant namespace on the cluster provider

> **You will need a similar namespace on the provider cluster** pointing to the cluster ID of your local cluster (cluster consumer).

Reviewer: I would shorten this text, as in the previous subsection.
I would also add that the namespaces on the two clusters do not need to have the same name.

For example, if the cluster consumer has cluster ID `cl-consumer`, its tenant namespace might look like the following:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    liqo.io/remote-cluster-id: cl-consumer
    liqo.io/tenant-namespace: "true"
  name: liqo-tenant-cl-consumer
spec: {}
```

## Declarative network configuration

By default, the network connection between clusters is established using a secure channel created via [Wireguard](https://www.wireguard.com/).

> To allow the creation of the tunnel, one cluster, typically the provider, needs to host a server gateway that exposes a port for the client gateway (usually running on the consumer cluster) to connect to.

Reviewer: ```suggestion
In this case, one cluster (usually the provider) needs to host a server gateway that exposes a UDP port that is reachable from the client gateway (usually running on the consumer cluster).
```

> Therefore, it is required that the server gateway is reachable from the peer cluster, to enable the creation of the tunnel between the two clusters.

Reviewer: I would remove this sentence. It has already been said before.

In this guide, **we will configure the client gateway on the consumer cluster and the server gateway on the provider cluster**, which is the most common setup.

> However, depending on your requirements, you may choose to reverse these roles.

Reviewer: ```suggestion
However, given that the setup of the network peering is independent from the offloading role of the cluster (i.e., consumer vs. provider), you may choose to invert the client/server roles in case this is more convenient for your setup.
```

The CIDRs are defined at installation time, you can check the values of the pod and the external CIDR configured in the local cluster by running:

```{code} bash
kubectl get networks.ipam.liqo.io -n liqo

NAME            DESIRED CIDR    REMAPPED CIDR   AGE
external-cidr   10.70.0.0/16    10.70.0.0/16    3h16m
internal-cidr   10.80.0.0/16    10.80.0.0/16    3h16m
pod-cidr        10.244.0.0/16   10.244.0.0/16   3h16m
service-cidr    10.96.0.0/16    10.96.0.0/16    3h16m
```

### Creating and exchanging the Wireguard keys (both clusters)

To enable authentication and encryption, Wireguard requires a key pair, one for each gateway.

Those keys can be created [via the Wireguard utility tool](https://www.wireguard.com/quickstart/#key-generation) or via OpenSSL like so:

```{code} bash
openssl genpkey -algorithm X25519 -outform der -out private.der
openssl pkey -inform der -in private.der -pubout -outform der -out public.der

# Get the Wireguard private key
echo "Private key:"; cat private.der | tail -c 32 | base64
# Get the Wireguard public key
echo "Public key:"; cat public.der | tail -c 32 | base64
```

> At this point, **in each cluster**, we will need to create a secret containing this pair of keys:

Reviewer: ```suggestion
At this point, you need to create **in each cluster** a secret containing this pair of keys:
```

Reviewer: Be uniform. If you use "we", we use "we" always. In any case, "you" is usually preferred to "we", and often we avoid using the future (hence "you need to" instead of "we will need to") to be more concise.

```{code} yaml
apiVersion: v1
kind: Secret
metadata:
  labels:
    liqo.io/remote-cluster-id: <REMOTE_CLUSTER_ID>
  name: gw-keys
  namespace: <TENANT_NAMESPACE>
type: Opaque
data:
  privateKey: <WIREGUARD_PRIVATE_KEY>
  publicKey: <WIREGUARD_PUBLIC_KEY>
```

where WIREGUARD_PRIVATE_KEY and WIREGUARD_PUBLIC_KEY are the previously generated key pairs for that specific cluster (**each cluster has its pair of keys**).

Additionally, each cluster should have a PublicKey resource **with the public key of the peer cluster** (cluster A has the public key of B and vice versa):

```{code} yaml
apiVersion: networking.liqo.io/v1beta1
kind: PublicKey
metadata:
  labels:
    liqo.io/remote-cluster-id: <HERE_THE_CLUSTER_ID_OF_PEER_CLUSTER>
    networking.liqo.io/gateway-resource: "true"
  name: gw-publickey
  namespace: <TENANT_NAMESPACE>
spec:
  publicKey: <REMOTE_WIREGUARD_PUBLIC_KEY>
```

In order to make things work, make sure that the PublicKey resource has the labels:

```{code} yaml
liqo.io/remote-cluster-id: <HERE_THE_CLUSTER_ID_OF_PEER_CLUSTER>
networking.liqo.io/gateway-resource: "true"
```

### Configuring the server gateway (cluster provider)

> As mentioned before, a Wireguard tunnel connects the clusters peered with Liqo.

Reviewer: ```suggestion
By default, a Wireguard tunnel connects the clusters peered with Liqo.
```

Reviewer: Try to avoid "as mentioned before" or similar. Is like saying "I know that you are not paying attention, so I repeat here what I said already before".  Which is not perceived as a good message.

> In this section we are going to configure the gateway server, where the client will connect to.

Reviewer: ```suggestion
This section shows how to configure the gateway server, where the client will connect to.
```

The `GatewayServer` resource describes the configuration of the gateway server, and should be applied on the cluster acting as server for the tunnel creation.
You can check [here](./inter-cluster-network.md#creation-of-a-gateway-server) an example of the `GatewayServer` CR.

When you create the `GatewayServer` resource, **make sure to specify the `secretRef`** pointing to the key pairs we created before.

> Note that under `.spec.endpoint` of the `GatewayServer` resource you can configure a fixed `nodePort` or `loadBalancerIP` (if supported by your provider) to have a deterministic port or ip address for the gateway, so that the configuration of the client that connects to it can be defined in advance.

Reviewer: ```suggestion
Note that under `.spec.endpoint` of the `GatewayServer` resource you can configure a fixed `nodePort` or `loadBalancerIP` (if supported by your provider) to have a precise UDP port or IP address for the gateway, so that the configuration of the client that connects to it can be defined in advance.
```

### Configuring the client gateway (cluster consumer)

The other cluster, in this case the consumer, needs to run the client gateway connecting to the service exposed by the cluster provider.

The `GatewayClient` resource describes the configuration of the gateway client, **it should contain the parameters** to connect to service exposed by the gateway server.

The following is an example of the `GatewayClient` CR to connect to the gateway server that we previously configured:

```{code} yaml
apiVersion: networking.liqo.io/v1beta1
kind: GatewayClient
metadata:
  creationTimestamp: null
  labels:
    liqo.io/remote-cluster-id: <PROVIDER_CLUSTER_ID>   # the remote cluster ID
  name: client
  namespace: <CONSUMER_TENANT_NAMESPACE>
spec:
  clientTemplateRef:
    apiVersion: networking.liqo.io/v1beta1
    kind: WgGatewayClientTemplate
    name: wireguard-client
    namespace: liqo
  secretRef:
    name: <WIREGUARD_KEYS_SECRET_NAME>
  endpoint:
    addresses:
    - <REMOTE_IP>
    port: 30742
    protocol: UDP
  mtu: 1340
```

Where:

- `PROVIDER_CLUSTER_ID` is the cluster ID of the provider, where the gateway server is running;

> - `CONSUMER_TENANT_NAMESPACE` is the tenant namespace on the cluster consumer, where, in thie case, we are configuring the gateway client;

Reviewer: ```suggestion
- `CONSUMER_TENANT_NAMESPACE` is the tenant namespace on the cluster consumer, where, in this case, we are configuring the gateway client;
```

> - `WIREGUARD_KEYS_SECRET_NAME` is the name of the secret with the Wireguard key paris we created before;

Reviewer: ```suggestion
- `WIREGUARD_KEYS_SECRET_NAME` is the name of the secret with the Wireguard key pairs we created before;
```

> - `REMOTE_IP`: is the IP address of one of the nodes of the cluster provider, as we configured a `NodePort` service. If the service was a `LoadBalancer` the ip would be the one of the load balancer.

Reviewer: ```suggestion
- `REMOTE_IP`: is the IP address of one of the nodes of the cluster provider, as we configured a `NodePort` service. If the service was a `LoadBalancer` the IP would be the one of the load balancer.
```

### Summary of network configuration

To sum up, to set up the network, **both clusters need**:

- a `Configuration` resource with the network configuration of the peer cluster
- a `Secret` containing the Wireguard public and private keys
- a `PublicKey` with the Wireguard public key **of the peer cluster**

> The **cluster provider will have a `GatewayServer`** resource

Reviewer: ```suggestion
The **cluster provider will have a `GatewayServer`** resource.
```

> The **cluster consumer a `GatewayClient` resource** connecting to the peer Gateway server.

Reviewer: ```suggestion
```

Once you applied all the required resources, the client should be able to connect to the server and create the tunnel.

You can get the `Connection` resource to check the status of the tunnel, as shown [here](./inter-cluster-network.md#connection-crds).

## Declarative configuration of clusters authentication

> In this section we are going to configure the authentication between the clusters, allowing the cluster consumer to ask for resources.

Reviewer: ```suggestion
This section shows how to configure the authentication between the clusters, allowing the cluster consumer to ask for resources.
```

When authentication is manually configured, **the user is in charge of providing the credentials with the permission required** by the cluster consumer to operate.

If your are not familiar with how authentication works in Kubernetes you can check [this documentation page](https://kubernetes.io/docs/reference/access-authn-authz/authentication/).
You can also check [here to know how to issue a certificate for a user](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#normal-user).

```{warning}
Note that with EKS authentication via client certificate [is not directly supported](https://aws.amazon.com/blogs/containers/managing-access-to-amazon-elastic-kubernetes-service-clusters-with-x-509-certificates/).
You can check [here](https://docs.aws.amazon.com/eks/latest/userguide/cluster-auth.html) how access control works in eks.
```

### Cluster consumer role binding (cluster provider)

> Once we created the credentials the consumer can work with, we will need to provide the minimum permission required by the consumer to operate.

Reviewer: ```suggestion
Once we created (in the provider cluster) the credentials the consumer can work with, we need to provide the minimum permission required by the consumer to operate.
```

Note that **the cluster consumer will never directly create workloads on the remote cluster** and, at this stage, it should have only the permissions to create the liqo resources to ask for the approval of a `ResourceSlice`.

> To do so, we will need to bind the newly created user to the `liqo-remote-controlpane` role.

Reviewer: controlpane or controlplane?

> To do so, we will need to bind the newly created user to the `liqo-remote-controlpane` role.

Reviewer: ```suggestion
To do so, we need to bind the newly created user to the `liqo-remote-controlplane` role.
```

This can be done **by creating the following `RoleBinding` resource in the tenant namespace of the cluster provider**:

```{code} yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  labels:
    liqo.io/remote-cluster-id: <CONSUMER_CLUSTER_ID>
  name: liqo-binding-liqo-remote-controlplane
  namespace: <TENANT_NAMESPACE>
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: liqo-remote-controlplane
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: <USER_COMMON_NAME>
```

> Where, when the user authenticate via certificate signed by the cluster CA, `USER_COMMON_NAME` is the CN field of the certificate.

Reviewer: ```suggestion
where, when the user authenticates via certificate signed by the cluster CA, `USER_COMMON_NAME` is the CN field of the certificate.
```

### Creation of a tenant for the cluster consumer (cluster provider)

On the provider side, to allow the authentication of a consumer, we will need to create a `Tenant` resource for it.

> This resource is useful to control the remote consumer (e.g. if we the provider wants to prevent a remote consumer to negotiate more resources, it can set the tenant condition to `Cordoned`, stopping any other resources negotiation).

Reviewer: ```suggestion
This resource is useful to control the remote consumer (e.g. if the provider would like to prevent a remote consumer to negotiate more resources, it can set the tenant condition to `Cordoned`, stopping any other resources negotiation).
```

Reviewer: "want" is considered unpolite. Better "would like".

> Note that, in the case of declarative configuration, there won't be any handshake between the clusters, so we will need to configure the tenant so that it accepts the `ResourceSlice` of the given consumer, even though no handshake occurred.

Reviewer: ```suggestion
Note that, in the case of declarative configuration, there will not be any handshake between the clusters, so we will need to configure the tenant so that it accepts the `ResourceSlice` of the given consumer, even though no handshake occurred.
```

This can be done setting `TolerateNoHandshake` as `authzPolicy` like in the following example:

```{code} yaml
apiVersion: authentication.liqo.io/v1beta1
kind: Tenant
metadata:
  labels:
    liqo.io/remote-cluster-id: <CONSUMER_CLUSTER_ID>
  name: my-tenant
  namespace: <TENANT_NAMESPACE>
spec:
  clusterID: <CONSUMER_CLUSTER_ID>
  authzPolicy: TolerateNoHandshake
  tenantCondition: Active
```

### Add the credentials on the cluster consumer (cluster consumer)

> The previously created credentials on the cluster provider, should be provided to the cluster consumer.

Reviewer: ```suggestion
The previously created credentials on the cluster provider should be given to the cluster consumer.
```

```text
 k logs -n liqo liqo-crd-replicator-68f9f55dfc-bc4l8 --tail 5

I1107 10:05:29.249872       1 crdReplicator-operator.go:94] Processing Secret "liqo-tenant-cl02/kubeconfig-controlplane-cl02"
I1107 10:05:29.254977       1 reflector.go:90] [cl02] Starting reflection towards remote cluster
I1107 10:05:29.255001       1 reflector.go:131] [cl02] Starting reflection of offloading.liqo.io/v1beta1, Resource=namespacemaps
I1107 10:05:29.255035       1 reflector.go:131] [cl02] Starting reflection of authentication.liqo.io/v1beta1, Resource=resourceslices
I1107 10:05:29.355741       1 reflector.go:163] [cl02] Reflection of authentication.liqo.io/v1beta1, Resource=resourceslices correctly started
```

### Summary of authentication configuration

To sum up, to set up the authentication, **on the cluster provider** you will need to:

- Create the credentials to be used by the consumer
- Bind the credentials to the `liqo-remote-controlplane` role
- Create a `Tenant` resource for the consumer

While, **on the consumer side**:

- Create a new secret containing the kubeconfig with the credentials to access to the cluster provider

## Declarative configuration of namespace offloading

> While offloading is independent from the network, which means that it is possible to negotiate resources and configure a namespace offloading without the inter-cluster network enabled, **a [working authentication configuration](#declarative-configuration-of-clusters-authentication) is a pre-requisite to enable offloading**.

Reviewer: Got it. Modified accordingly.

### Ask for resources: configure a ResourceSlice

> The `ResourceSlice` resource is the one to be created to ask a cluster provider for resources.

Reviewer: ```suggestion
The `ResourceSlice` resource is the CR that defines the resources requested to the provider cluster.
```

Reviewer: Not clear where the ResourceSlice has to be created: consumer or provider cluster?

The following is an example of `ResourceSlice`:

```{code} yaml
apiVersion: authentication.liqo.io/v1beta1
kind: ResourceSlice
metadata:
  annotations:
    liqo.io/create-virtual-node: "true"
  creationTimestamp: null
  labels:
    liqo.io/remote-cluster-id: <PROVIDER_CLUSTER_ID>
    liqo.io/remoteID: <PROVIDER_CLUSTER_ID>
  name: test
  namespace: liqo-tenant-cl02
spec:
  class: default
  providerClusterID: <PROVIDER_CLUSTER_ID>
  resources:
    cpu: 20
    ram  128Gi
```

> With the configuration above, once the resources are accepted from the provider side, a new (virtual) node, impersonating the provider cluster, will make available the requested resorces on the consumer cluster.

Reviewer: ```suggestion
The above request has to be accepted by the provider cluster.
If the request is successful, a new (virtual) node, impersonating the provider cluster, will make available the requested resources on the consumer cluster.
```

Author: I changed a little bit the order of sentences to address both this comment and the one above

To know more about `ResourceSlice` and `VirtualNode` check [this section of the documentation](./offloading-in-depth.md#create-resourceslice).

### Enable offloading and K8s resources availability on remote clusters

By default, the virtual nodes are not eligible for task scheduling, unless offloading is enabled for the namespace where pod is running.
To do so, a `NamespaceOffloading`, like the following, should be created:

```{code} yaml
apiVersion: offloading.liqo.io/v1beta1
kind: NamespaceOffloading
metadata:
  name: offloading
  namespace: demo
spec:
  clusterSelector:
    nodeSelectorTerms: []
  namespaceMappingStrategy: DefaultName
  podOffloadingStrategy: LocalAndRemote
```

> The `NamespaceOffloading` resource should be created in the namespace that we want to extend on the remote clusters.

Reviewer: ```suggestion
The `NamespaceOffloading` resource should be created in the namespace that we would like to extend on the remote clusters.
```

For example, the resource above, extends the `demo` namespace on all the configured provider clusters.

> As the policy is `LocalAndRemote` the pod will be executed locally or remotely and it depends on the choice made by the the Kubernetes scheduler.

Reviewer: ```suggestion
Since the `podOffloadingStrategy` policy is `LocalAndRemote`, the a new pod could be executed either locally or remotely depending on the choice made by the vanilla Kubernetes scheduler (e.g., if the remote virtual node is plenty of free resources, it may be preferred against local nodes that are already used by other pods).
```

[Check here](../../usage/namespace-offloading.md#namespace-mapping-strategy) to know more about namespace offloading.

```{warning}

> Currently, **there is a caveat with the namespace offloading preventing the pods created before the creation of the `NamespaceOffloading` resource to be scheduled on a remote cluster**.

Reviewer: ```suggestion
Currently, the `NamespaceOffloading` resource **must be created before scheduling a pod on a remote cluster**.
```

> For example, if we configure a pod to run on the remote cluster, but pod is created before the `NamespaceOffloading`, that pod remains in a `Pending` state, also after the namespace is actually offloaded.

Reviewer: ```suggestion
For example, if we configure a pod to run on the remote cluster, but the pod is created before setting the `NamespaceOffloading` resource, that pod will remain in a `Pending` state forever, also after the namespace is actually offloaded.
```

In this case, one cluster (usually the provider) needs to host a server gateway that exposes a UDP port that must be reachable from the client gateway (usually running on the consumer cluster).

In this guide, **we will configure the client gateway on the consumer cluster and the server gateway on the provider cluster**, which is the most common setup.

However, given that the setup of the network peering is independent from the offloading role of the cluster (i.e., consumer vs. provider), you may choose to invert the client/server roles in case this is more convenient for your setup.

### Creating and exchanging the network configurations (both clusters)

The clusters that needs to be connected requires the network configuration of the peer cluster, which is provided via the `Configuration` CR.
You can check an example of the resources to apply at the [following documentation page](./inter-cluster-network.md#definition-of-the-network-configuration-configuration-crds).

**The Configuration resource should be applied in both clusters** and it **must contain the pod and external CIDR of the peer cluster**. (cluster A has the CIDR config of B and vice versa).

> The CIDRs are defined at installation time, you can check the values of the pod and the external CIDR configured in the local cluster by running:

Author: Probably we can change the command with something like this:
```
kubectl get networks.ipam.liqo.io -n liqo external-cidr -o=jsonpath={'.status.cidr'}
```
which directly returns the network to use

While offloading is independent from the network, which means that it is possible to negotiate resources and configure a namespace offloading without the inter-cluster network enabled, **a [working authentication configuration](#declarative-configuration-of-clusters-authentication) is a pre-requisite to enable offloading**.

### Ask for resources: configure a ResourceSlice (consumer cluster)

The `ResourceSlice` resource is the CR that defines the computational resources requested by the consumer to the provider cluster.
It should be created on the tenant namespace of the consumer cluster, and it is automatically forwarded to the provider cluster, which can accept or reject it.

The following is an example of `ResourceSlice`:

```yaml
apiVersion: authentication.liqo.io/v1beta1
kind: ResourceSlice
metadata:
  annotations:
    liqo.io/create-virtual-node: "true"
  creationTimestamp: null
  labels:
    liqo.io/remote-cluster-id: <PROVIDER_CLUSTER_ID>
    liqo.io/remoteID: <PROVIDER_CLUSTER_ID>
  name: test
  namespace: <CONSUMER_TENANT_NAMESPACE>
spec:
  class: default
  providerClusterID: <PROVIDER_CLUSTER_ID>
  resources:
    cpu: 20

>     ram  128Gi

Reviewer 2: Hello @claudiolor .

I suggest the following modifications on this yaml:

resources:
cpu: '20'
ram : 128Gi
pods: '110'

As reported into this slack [thread](https://liqo-io.slack.com/archives/C014Z7F25KP/p1732027216878499).

Sorry if I'm just commenting, but I don't have permissions for editing files.

Thanks.

Author: Hi Francesco, if the `pods` resource is missing, then the default value is used, so I would not add it in this part of the documentation for simplicity.
If you are experiencing any issue when you don't provide the resources, I suggest to open an issue in this repository with the details of your problem and how to reproduce it, if it is actually a bug, then we will take care of fixing it.

Author: Ops, thanks for let me notice it :)
